diff-tests/core/model/src/model.ts

```diff
@@ -0,0 +1,671 @@
+      });
+      this.vscIDtoOpIDs.delete(data.vscID);
+    }
+  };
+
+  onReceiveSlash = (data: Slash) => {
+    let infractionHeight = undefined;
+
+    if (data.vscID === 0) {
+      infractionHeight = this.initialHeight;
+    } else {
+      infractionHeight = this.vscIDtoH[data.vscID];
+    }
+
+    if (this.m.staking.status[data.val] === Status.UNBONDED) {
+      return;
+    }
+
+    if (data.isDowntime) {
+      this.m.events.push(Event.RECEIVE_DOWNTIME_SLASH_REQUEST);
+    } else {
+      this.m.events.push(Event.RECEIVE_DOUBLE_SIGN_SLASH_REQUEST);
+    }
+
+    if (this.tombstoned[data.val]) {
+      return;
+    }
+
+    this.m.staking.slash(data.val, infractionHeight);
+    this.m.staking.jailUntil(data.val, this.m.t[P] + JAIL_SECONDS);
+    if (data.isDowntime) {
+      this.downtimeSlashAcks.push(data.val);
+    } else {
+      this.tombstoned[data.val] = true;
+    }
+  };
+
+  afterUnbondingInitiated = (opID: number) => {
+    if (!this.vscIDtoOpIDs.has(this.vscID)) {
+      this.vscIDtoOpIDs.set(this.vscID, []);
+    }
+    this.vscIDtoOpIDs.get(this.vscID)!.push(opID);
+  };
+}
+
+class CCVConsumer {
+  m;
+  // maps consumer height h to the id of the last vscid
+  // received at height h-1
+  hToVscID: Record<number, number>;
+  // validator power changes pending aggregation
+  pendingChanges: Record<Validator, number>[];
+  // maps vscid to earliest timestamp to mature
+  maturingVscs: Map<number, number>;
+  // is there an outstanding downtime operation for a validator?
+  outstandingDowntime: boolean[];
+  // array of validators to power
+  // value undefined if validator is not known to consumer
+  power: (number | undefined)[];
+
+  constructor(model: Model, { ccvC }: ModelInitState) {
+    this.m = model;
+    Object.assign(this, ccvC);
+  }
+
+  beginBlock = () => {
+    this.hToVscID[this.m.h[C] + 1] = this.hToVscID[this.m.h[C]];
+  };
+
+  endBlock = () => {
+    // Unbond all the matured VSCs
+    const matured = (() => {
+      const ret: number[] = [];
+      this.maturingVscs.forEach((time, vscID) => {
+        if (time <= this.m.t[C]) {
+          ret.push(vscID);
+        }
+      });
+      return ret;
+    })();
+    matured.forEach((vscID) => {
+      const data: VscMatured = { vscID };
+      this.m.events.push(Event.CONSUMER_SEND_MATURATION);
+      this.m.outbox[C].add(data);
+      this.maturingVscs.delete(vscID);
+    });
+    // gather and apply any changes
+    if (this.pendingChanges.length < 1) {
+      return;
+    }
+    const changes = (() => {
+      const ret: Map<Validator, number> = new Map();
+      this.pendingChanges.forEach((updates) => {
+        Object.entries(updates).forEach(([val, power]) =>
+          ret.set(parseInt(val), power),
+        );
+      });
+      return ret;
+    })();
+
+    this.pendingChanges = [];
+
+    changes.forEach((power, val) => {
+      this.power[val] = undefined;
+      if (0 < power) {
+        this.m.events.push(Event.CONSUMER_UPDATE_POWER_POSITIVE);
```

Comment:
Add event for adding a new validator. Basically, there are three events: adding a new one, updating an existing one with a new power, removing am existing one.